File: solutions/ts/17.ts
function detectBombs(grid: boolean[][]): number[][] {
  const minesGrid: number[][] = [];
  for (let i = 0; i < grid.length; i++) {
    minesGrid.push([]);
    for (let j = 0; j < grid[i].length; j++) {
      minesGrid[i].push(0);
      let counter = 0;
      for (let x = -1; x <= 1; x++) {
        for (let y = -1; y <= 1; y++) {
          if (x === 0 && y === 0) continue; // skip the current cell
          if (grid[i + x]?.[j + y]) counter++;
        }
      }
      minesGrid[i][j] = counter;
    }
  }
  return minesGrid;
}

const bombs1 = detectBombs([
  [true, false, false],
  [false, true, false],
  [false, false, false],
]);
console.log(bombs1);
// [
//   [1, 2, 1],
//   [2, 1, 1],
//   [1, 1, 1]
// ]

const bombs2 = detectBombs([
  [true, false],
  [false, false],
]);
console.log(bombs2);
// [
//   [0, 1],
//   [1, 1]
// ]

const bombs3 = detectBombs([
  [true, true],
  [false, false],
  [true, true],
]);
console.log(bombs3);
// [
//   [1, 1],
//   [4, 4],
//   [1, 1]
// ]

/* 
  Score: ★★★★★
*/

File: solutions/python/17.py
def detect_bombs(grid: list[list[bool]]) -> list[list[int]]:
    mines_grid = []
    i = 0
    while (i < len(grid)):
        mines_grid.append([])
        j = 0
        while (j < len(grid[i])):
            mines_grid[i].append(0)
            counter = 0
            for x in [-1, 0, 1]:
                for y in [-1, 0, 1]:
                    row_index = i+x
                    cell_index = j+y
                    if (x == 0 and y == 0):
                        continue
                    if (
                        row_index >= 0 and
                        row_index < len(grid) and
                        cell_index >= 0 and
                        cell_index < len(grid[i]) and
                        grid[row_index][cell_index]
                    ):
                        counter += 1
            mines_grid[i][j] = counter
            j += 1
        i += 1
    return mines_grid


if __name__ == "__main__":
    bombs1 = detect_bombs([
        [True, False, False],
        [False, True, False],
        [False, False, False],
    ])
    print(bombs1)
    # [
    #   [1, 2, 1],
    #   [2, 1, 1],
    #   [1, 1, 1]
    # ]

    bombs2 = detect_bombs([
        [True, False],
        [False, False],
    ])
    print(bombs2)
    # [
    #   [0, 1],
    #   [1, 1]
    # ]

    bombs3 = detect_bombs([
        [True, True],
        [False, False],
        [True, True],
    ])
    print(bombs3)
    # [
    #   [1, 1],
    #   [4, 4],
    #   [1, 1]
    # ]

"""
  Score: ★★★★★
"""
